### ISSUE TITLE
wasm-emscripten-finalize fails without error message on 2GB wasm file

### ISSUE THREAD

**Issue by Markus87** [emscripten-core/emsdk#867]
emsdk 2.0.26 / Windows 10

The command it fails on is: 
`wasm-emscripten-finalize --minimize-wasm-changes -g --dyncalls-i64 --dwarf Big.wasm -o Big.wasm --detect-features`

If I remove `--dwarf` there is no error, a new smaller wasm file is output.

Currently trying to let it run with `--debug`, but I dont think it will finish in the time given to me in this world.
Is there anything else I can try to help me figure out whats the problem?

If requested I can provide the Big wasm file.

**Comment by kripken** [comment]
I suspect this may be a case of running out of memory - DWARF rewriting uses more of it. Watching memory usage may help confirm that. You can also run in a debugger and get a stack trace of where it crashes.

**Comment by Markus87** [comment]
Yeah, that is was what I suspected. I watched the process go up to 53GB. I have 64GB of memory installed and I had a 100GB pagefile, so memory should not have been a problem. 🤔 

But attaching the debugger is worth a try. 👍

**Comment by Markus87** [comment]
With the attached debugger I see this exception:
![grafik](https://user-images.githubusercontent.com/6339979/129718800-5301f33d-26b4-4fc6-a84c-32eabf131e9e.png)

It occured way after the peak memory usage, the process used about 13GB of memory at that point.
Are the debugsymbols available for Windows somewhere?
Or should I upload the wasm file which reproduces it?

**Comment by kripken** [comment]
As it's an OOM, maybe there isn't much to do here. That much memory sounds excessive, but as this is a 2GB wasm file, maybe that is not surprising - most of that file will be DWARF, and DWARF is quite compact in binary form, so it expands into a very large amount of memory when it is being updated.

I think for this file you may need to build without DWARF debug info. Or you can try things like `-gline-table-only` which limit the type of debug info.

**Comment by Markus87** [comment]
Thanks for the reponse!

I am confused, how do you see that this crash is OOM?
When the process throws the exception it uses about 20% of my installed physical memory.
If this really is just a case of OOM I will happily close this issue. 😬 

I understand having the need for such big files seems a little crazy, but I am porting a big desktop application with many plugins that need to be linked whole in order to work.

If it helps in any way you can contact me via email(`development@bitfactory.at`) and I will provide you with a link to the file in question.

Thank you for the possible workaround `-gline-table-only`, I will check if it helps in this case.

**Comment by kripken** [comment]
Well, I'm not sure, but after reading this I'd put money on it being an OOM :smile: 

> I watched the process go up to 53GB.

There might be a quick spike that you don't see right before the OS kills it, or maybe the OS some other internal limit you are hitting, even if you do have more memory available.

Do you actually need DWARF here, btw? Perhaps source maps would be enough, if line tables are. Basically, using less or no debug info sounds like the solution to me.

**Comment by Markus87** [comment]
DWARF was the only thing I got working in the browser for debugging, also to me it read in the docs like DWARF will be the future of debugging for WebAssembly.
Other than that I had no reason for it specifically.

Since you are so sure that this is an OOM issue I am going to close the issue.
You also provided me with possible work arounds that may work for me.

For now I limited the amount of libraries that are linked to prevent the crash. (makes it look like a resource problem even more)
If I should find out that it crashes because of a specific reason that I can reproduce I may reopen this issue or make a new one.

Thanks for your help!
